Refactor GOL state to a new object, `GolWorld`

vijoc opened the pull request:
For now, performance is roughly equal to what it was. Performance
optimization can be done later on with smarter neighbour counting and
state storage.

* Move state responsibility from `GolScene` to `GolWorld`
* Introduce concept of neighbourhoods; may be modified later on
* Introduce `CellContainer` and methods for modifying it; for now, just
  a `std::vector<std::vector<bool>>`.
* Miscellaneous small fixes & cleanups

vijoc committed:
Refactor GOL state to a new object, `GolWorld`

For now, performance is roughly equal to what it was. Performance
optimization can be done later on with smarter neighbour counting and
state storage.

* Move state responsibility from `GolScene` to `GolWorld`
* Introduce concept of neighbourhoods; may be modified later on
* Introduce `CellContainer` and methods for modifying it; for now, just
  a `std::vector<std::vector<bool>>`.
* Miscellaneous small fixes & cleanups

jukkae commented:
Coordinate wrapping, neighbor counting and / or world resizing is buggy: With some rule sets (especially fast-growing ones), when zoomed in a bit, clearing the world with C clears the visible area, but even without reseeding, live cells will soon appear from the borders. Haven't looked into the code yet, but I assume that when zooming in, the neighborhoods must be recalculated for cells on the boundary.

vijoc committed:
Fix issue with ghost cells after clear

Row wrapping was incorrect, which (for some reason) caused cell states
to remain after clear in the bottom row. Could not replicate the bug
after this fix.

Also removed obsolete check in step iteration.

vijoc commented:
Issue should be fixed now; will update with some final cleanups of code later today.

vijoc committed:
Clean up `GolWorld`

* Add some documentation to `GolWorld.hpp`.
* Extract `wrapAround` to a new file under `src/util/`.
* Reorder implementations in `GolWorld.cpp`.
* Miscellaneous small touches.

vijoc commented:
Cleaned up the implementation. There is still room for further refactoring, and no optimizations were yet implemented. However, this PR should suffice to make further modifications simpler and more focused.
